### DISCUSSION TITLE
Downloaded MicroPython to stm32h7b3i-dk board, but not able to save code

### DISCUSSION THREAD

**Discussion by 4riceotter** [micropython/micropython#11584] [STM32 / Pyboard] [answered]
Hi, 

    I just got Micropython [v1.20.0 (2023-04-26)] downloaded onto the stm32h7b3i-dk board using the .hex code from:

        https://micropython.org/download/STM32H7B3I_DK/

with stm32 programmer on Windows.  I fired up Thonny (3.3.13) editor (which I've been using for RPi Pico uPython work) and was able to select Generic Micropython.  REPL started with this line:

        MicroPython v1.20.0 on 2023-04-26; STM32H7B3I-DK with STM32H7B3LIH6Q

but I couldn't save my source file (like blinky.py) to the STM32H7B3I_DK.  Has anyone run into this issue before, and does a solution already exist to get it working?  Also, do I need some kind of BSP file to get it to work?

Much appreciation for any pointer you can provide!

Benjamin

**Comment by jimmo** [comment]
If you have a REPL then that should be all you need to copy files. How are you connected to the board, is it via the ST-Link UART or via the USB VCP?

> but I couldn't save my source file (like blinky.py) to the STM32H7B3I_DK

What do you mean by "couldn't save". What goes wrong? 

> Also, do I need some kind of BSP file to get it to work?

The MicroPython-equivalent of the BSP is what generated that firmware image. i.e. https://github.com/micropython/micropython/tree/master/ports/stm32/boards/STM32H7B3I_DK

Thonny doesn't need any board-specific logic.

Either way, can you try copying files using `mpremote` instead of Thonny and see if that works?

**Reply by 4riceotter** [reply]
Hi, Jim,

    Thanks for the quick reply!  When I tried to save my python code I got the following error:

![FileSaving_Error](https://github.com/micropython/micropython/assets/101532401/f443b745-d143-44e1-b0d7-86813c984279)

I installed mpremote (1.20.0) on Windows 10 using "pip install mpremote", and it was able to find and connect to the stm32 H7 board through COM8.  My attempt to copy is listed below:

C:\Users\riceo\Documents\STMicro\MicroPython>mpremote mount . cp blinky.py /remote
Local directory . is mounted at /remote
cp blinky.py /remote
mpremote: [Errno 13] Permission denied: '/remote'

I wonder how I can get this write permission?   As a try I launched bonjour (the file sharing wizard from Apple, as somewhere I learned that Windows requires that to allow access to USB, but it made no difference).

Benjamin

**Reply by jimmo** [reply]
Sorry I don't know how to advise on Thonny.

With `mpremote` I think you might be confusing the purpose of the `mount` command.

`mpremote mount` makes the local path (in this case, `.`, the current directory) available as `/remote` on the device.
When you subsequently do `cp blinky.py /remote` that is trying to copy the file to`/remote` on your computer. You need to use `:` to indicate a remote path (on the device). (This works the same way as the `scp` tool). This is failing because you aren't allowed to write to a path `/remote` on your Mac.

But you don't want to copy it to `/remote` on the device, because that's actually the current directory on your PC (via the mount).

Instead, if you just want to copy a file to your device, use: `mpremote cp blinky.py :blinky.py` (or `mpremote cp blinky.py :` where the destination file name is implicit from the source filename). (i.e. do not use mount for copying files).

If you want to use `mount mount`, then theres no need to copy the file. Just run `mpremote mount .` then at the REPL you can do `import blinky`. (This will import the file from the mounted path, which is actually behind the scenes mapped to your PC -- this is really useful while you're developing code because you don't have to keep copying files).

**Reply by jimmo** [reply]
FWIW we understand the documentation for mpremote needs a lot of work, especially to describe how the different features can be used in conjunction with eachother. We're also planning to make some introductory videos at some stage... but as always there's so many things to do!

**Reply by 4riceotter** [reply]
Hi, Jim,

    Thanks for the explanation.  I was wondering about the syntax to use mpremote with, but I thought a blind stab would at least take me one step closer ;-)

    So here's the result with "mpremote cp blinky.py :blinky.py", and it gave the same error as with thonny:

cp blinky.py :blinky.py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 19] ENODEV

It seems to be an issue with Windows ("OSError") not allowing access to the board.  My next step is to try rshell:

        Successfully installed pyreadline-2.1 pyudev-0.24.1 rshell-0.0.31

but the command itself gave out the following error:

C:\Users\riceo\Documents\STMicro\MicroPython>rshell
Traceback (most recent call last):
  File "C:\Python310\Scripts\rshell-script.py", line 33, in <module>
    sys.exit(load_entry_point('rshell==0.0.31', 'console_scripts', 'rshell')())
  File "C:\Python310\Scripts\rshell-script.py", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "C:\Python310\lib\importlib\metadata\__init__.py", line 162, in load
    module = import_module(match.group('module'))
  File "C:\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Python310\lib\site-packages\rshell\command_line.py", line 1, in <module>
    import rshell.main
  File "C:\Python310\lib\site-packages\rshell\main.py", line 80, in <module>
    import readline
  File "C:\Python310\lib\site-packages\readline.py", line 34, in <module>
    rl = Readline()
  File "C:\Python310\lib\site-packages\pyreadline\rlmain.py", line 422, in __init__
    BaseReadline.__init__(self)
  File "C:\Python310\lib\site-packages\pyreadline\rlmain.py", line 62, in __init__
    mode.init_editing_mode(None)
  File "C:\Python310\lib\site-packages\pyreadline\modes\emacs.py", line 633, in init_editing_mode
    self._bind_key('space',       self.self_insert)
  File "C:\Python310\lib\site-packages\pyreadline\modes\basemode.py", line 162, in _bind_key
    if not callable(func):
  File "C:\Python310\lib\site-packages\pyreadline\py3k_compat.py", line 8, in callable
    return isinstance(x, collections.Callable)
AttributeError: module 'collections' has no attribute 'Callable'

   I'll try the whole process again on my Fedora 37 Linux system, and I'll post another update then.

   Thanks again for providing the great help!

Benjamin

**Reply by jimmo** [reply]
OHHH. Sorry I missed this initially @4riceotter -- the MicroPython board definition (BSP) for the STM32H7B3I-DK doesn't enable the internal filesystem...

It's saying ENODEV, because, there is no flash device for the filesystem.

I don't know what's going on with this board definition. I'm not sure why it doesn't enable either the internal flash or the external 512Mbit octo-spiflash. My guess is that the driver doesn't support octo-spiflash properly, but disabling the internal flash was an oversight.

If you want to try building your own firmware, then I think the best thing to do is to change

```
#define MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE (0)
```
to
```
#define MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE (1)
```

There's also an issue with the linker script. I will send a PR. Anyway, I've attached .hex and .dfu files which are v1.20 with those changes to enable the internal filesystem. Please let me know how that goes.
[firmware.zip](https://github.com/micropython/micropython/files/11526847/firmware.zip)

**Reply by jimmo** [reply]
See https://github.com/micropython/micropython/pull/11585

**Reply by 4riceotter** [reply]
Hi, Jim,

    Thank you for the update.  I unzipped the firmware.zip file and tried copying the two files to the stm32-H7 board, as well as use the STM32 Programmer to flash the .hex file onto the board.  The STM32Cube Programmer claimed that download is successful, but in either case REPL did not come up in thonny:

![image](https://github.com/micropython/micropython/assets/101532401/919a3044-9188-4a76-a885-96b689731bc8)

![image](https://github.com/micropython/micropython/assets/101532401/f3238808-ee37-4652-bbfe-445a2729241a)

no matter how many times I power cycle, clicked Stop, or press the board RESET button.   Mpremote claims that it's connected to the board, but trying any command to access the file system simply hangs the shell:

C:\Users\riceo>mpremote
Connected to MicroPython at COM8
Use Ctrl-] or Ctrl-x to exit this shell

C:\Users\riceo>mpremote mount .

Sorry, seems that I'd have to bother you some more ...

Benjamin

**Reply by jimmo** [reply]
> tried copying the two files to the stm32-H7 board,

Assume you mean copying to the st-link MSC drive. This should only be done with the .hex file. (It'll ignore the .dfu file I hope). I just included the .dfu file in case you were programming with a dfu programmer.

Was going to suggest doing a full erase first, but looks like cube programmer did a full chip erase first, so don't think that's the issue.

Unfortunately I don't have one of these boards... Perhaps there's some other reason the internal flash was disabled.

**Reply by jimmo** [reply]
@4riceotter I found the issue. Please see the updated .hex file which should now work to enable the internal filesystem (see updated PR https://github.com/micropython/micropython/pull/11585).

firmware.zip (removed, see below)

I also spoke to @dpgeorge we're going to also make it possible to use the external spiflash on this board (which you can enable via a config option at build time).

**Reply by jimmo** [reply]
Argh missed something else. The H7B3 has 8kiB sectors. Here's another updated firmware.hex.
[firmware.zip](https://github.com/micropython/micropython/files/11528149/firmware.zip)

**Reply by jimmo** [reply]
(I suspect there are likely to be more issues. It's a bit pointless without an actual board to test on. @dpgeorge has one, so I'll leave it with him to continue from here.)

**Reply by 4riceotter** [reply]
@jimmo @dpgeorge Thank you both for the hard work.  REPL still has not come back with any of the firmware.hex (including the one from 9 min. ago).  Still, appreciate all the time you have spent on this issue.  I'll wait for the good news from Damien then :-).  Meanwhile, I'll play with the original version on the micropython site (with which REPL comes up) to play with the other features of the board.

Benjamin
Benjamin

**Reply by jimmo** [reply, marked as answer]
@4riceotter here's a firmware.hex (from @dpgeorge) that will use the external spiflash (in regular, not octal mode). But should be enough to get you a working filesystem.
[firmware.zip](https://github.com/micropython/micropython/files/11528381/firmware.zip)

I don't know when we'll get around to supporting the internal flash on this chip properly (the H7B has some extra quirks over the the rest of the H7 family). It's quite similar to what we need to do for the H5, so probably follow https://github.com/micropython/micropython/pull/11138 for updates.

**Reply by 4riceotter** [reply]
It works!  You guys are wonderful @jimmo @dpgeorge :-)!

Thank you, thank you!

Benjamin
